### PR TITLE
#10108 add support for events in interfaces

### PR DESCRIPTION
**Pull request by omnilight**
Adds support for events in interfaces #10108

### CODE REVIEW

**Comment by SilverFire** [comment]
I like this idea! @samdark said in the [comment](https://github.com/yiisoft/yii2/issues/10108#issuecomment-154513513) that PR can drop speed of execution, but the code looks simple. Can you make a benchmark to check the difference?

**Comment by omnilight** [comment]
Ok, I've tested my code with the following script:

```php
namespace app\commands;
use yii\base\Component;
use yii\base\Event;
use yii\console\Controller;


/**
 * Class SpeedTestController
 */
class SpeedTestController extends Controller
{
    public $counter = 0;

    public function actionIndex()
    {
        Event::on(SomeClass::class, SomeInterface::EVENT_FROM_INTERFACE, function (Event $e) {
            $this->counter++;
        });

        $class = new SomeClass();

        $start = microtime(true);
        for ($i = 0; $i < 10000; $i++) {
            $class->emitEvent();
        }
        $end = microtime(true);

        echo $end - $start;
    }
}

interface SomeInterface
{
    const EVENT_FROM_INTERFACE = 'fromInterface';
}

class SomeClass extends Component implements SomeInterface
{
    public function emitEvent()
    {
        $this->trigger(self::EVENT_FROM_INTERFACE);
    }
}
```

Results:

Code | Result
-------  | ---------
Original | 0.4721
With pull-request | 1.45

I noticed, that `call_user_func` takes most of the time, so I've refactored code, after that:

Code | Result
-------  | ---------
Refactoring | 0.51

A bit slower, but not to much

**Comment by omnilight** [comment]
It is also possible to cache results of the `class_implements`, but it will increase amount of memory used

**Comment by samdark** [comment]
0.4721 to 0.51 for 10000 events isn't much but... how it behaves if there are multiple handlers?

**Comment by omnilight** [comment]
```php
namespace app\commands;
use yii\base\Component;
use yii\base\Event;
use yii\console\Controller;


/**
 * Class SpeedTestController
 */
class SpeedTestController extends Controller
{
    public $counter = 0;

    public function actionIndex()
    {
        for ($i = 0; $i < 100; $i++) {
            Event::on(SomeClass::class, 'event_'.$i, function (Event $e) {
                $this->counter++;
            });
        }

        $class = new SomeClass();

        $start = microtime(true);
        for ($i = 0; $i < 10000; $i++) {
            $class->emitEvent();
        }
        $end = microtime(true);

        echo $end - $start;
    }
}

interface SomeInterface
{

}

class SomeClass extends Component implements SomeInterface
{
    public function emitEvent()
    {
        for ($i = 0; $i < 100; $i++) {
            $this->trigger('event_'.$i);
        }
    }
}
```

Results:

Code | Result
-------- | ---------
Original | 45.301
Pull-request | 49.247

But if some class implements a lot of interfaces, it could be worth

**Comment by cebe** [comment]
> But if some class implements a lot of interfaces, it could be worth

that is not a common case so that should be fine.

**Comment by omnilight** [comment]
Oh, I've noticed, that @samdark meant "multiple handlers", ok, this code:

```php
<?php

namespace app\commands;

use yii\base\Component;
use yii\base\Event;
use yii\console\Controller;


interface SomeInterface
{

}

/**
 * Class SpeedTestController
 */
class SpeedTestController extends Controller
{
    public $counter = 0;

    public function actionIndex()
    {
        for ($i = 0; $i < 100; $i++) {
            Event::on(SomeClass::class, 'event', function (Event $e) {
                $this->counter++;
            });
        }

        $class = new SomeClass();

        $start = microtime(true);
        for ($i = 0; $i < 10000; $i++) {
            $class->emitEvent();
        }
        $end = microtime(true);

        echo $end - $start;
    }
}

class SomeClass extends Component implements SomeInterface
{
    public function emitEvent()
    {
        $this->trigger('event');
    }
}
```

will give:

Code | Result
-------- | --------
Original | 19.31
Pull | 19.50

**Comment by samdark** [comment]
Looks acceptable. @cebe, @klimov-paul what do you think?

File: framework/base/Event.php [hunk 1]
```diff
@@ -192,5 +203,17 @@ public static function trigger($class, $name, $event = null)
                 }
             }
         } while (($class = get_parent_class($class)) !== false);
```
Comment (klimov-paul) [inline framework/base/Event.php]: Why don't use `class_parents()` here then?
Both cycles can be unified this way.

**Comment by klimov-paul** [comment]
Perhaps it could be better to use `ReflectionClass` for both parent class and interface checks.

File: framework/base/Event.php (hunk 1 above)
Comment (omnilight) [inline framework/base/Event.php]: In original algorithm check for event listeners is performed in the following way:

1. Listeners for $class
2. Listeners for $class's parents
3. Listeners for $class's parents's parentents
4. So on...

If we will use `class_parents()` an order will be unpredictable

**Comment by omnilight** [comment]
Will check for perfomance in that case...

**Comment by SilverFire** [comment]
@omnilight any news?

**Comment by omnilight** [comment]
I'll work on this current weekend, and will update current issue

**Comment by omnilight** [comment]
Ok, I've tested reflection implementation of the event class (when we simply replaces calls to class_parent and class_implements to their reflection equivalent):

Version | Speed
---------- | ---------
Original | 18.72
Reflection | 19.28

So reflection is slower, and there is no reason to use it (and it does not give us any improvements on the algorithm of searching for class parents and it's interfaces)

I've updated the pull-request to use class_parents, as suggested by @klimov-paul - as I can understand this method parents in the correct order

**Comment by samdark** [comment]
Now it looks good to me. Need docs though.

**Comment by omnilight** [comment]
Ok, what kind of docs I should provide?

File: framework/base/Event.php [hunk 2]
```diff
@@ -24,6 +24,7 @@
  */
 class Event extends Object
 {
+    private static $_events = [];
```
Comment (SilverFire) [inline framework/base/Event.php]: We need a PHPDoc here

File: framework/base/Event.php [hunk 3]
```diff
@@ -145,11 +143,18 @@ public static function hasHandlers($class, $name)
         } else {
             $class = ltrim($class, '\\');
         }
-        do {
+
+        $classes = array_merge(
```
Comment (SilverFire) [inline framework/base/Event.php]: We need to add update PHPDoc for `hasHandlers` method and mention new feature there

**Comment by SilverFire** [comment]
@omnilight there is a guide for [Events](http://www.yiiframework.com/doc-2.0/guide-concept-events.html) and you feature should be mentioned there as well.

**Comment by omnilight** [comment]
I'll try to write docs, but I'll need help with english :)

**Comment by SilverFire** [comment]
@omnilight можешь написать доку на русском, а я переведу на английский

**Comment by omnilight** [comment]
@SilverFire Спасибо, напишу :)

**Comment by omnilight** [comment]
Added docs in russian

File: docs/guide-ru/concept-events.md [hunk 4]
```diff
@@ -222,6 +222,65 @@ Event::off(Foo::className(), Foo::EVENT_HELLO, $handler);
 Event::off(Foo::className(), Foo::EVENT_HELLO);
 ```
 
+Обработчики событий на уровне интерфейсов <span id="interface-level-event-handlers"></span>
+-------------
+
+Еще более общим случаем, чем обработка событий на уровне класса, является обработка на уровнне *интерфейсов*. Для этого по аналогии с обработчиками событий на уровне классов, необходимо вызвать статический метод [[yii\base\Event::on()]], передав ему в качестве первого параметра имя интерфейса. В этом случае инициализация события для любого класса, реализующего указанный интерфейс, будет приводить к срабатыванию соответствующего обработчика.
```
Comment (SilverFire) [inline docs/guide-ru/concept-events.md]: `Для этого` -> `Для этого, `

**Comment by SilverFire** [comment]
How about https://gist.github.com/SilverFire/467b2def3115ecc38c43? :)

**Comment by omnilight** [comment]
Greate!, I'll include englush version and will change russian variant

**Comment by SilverFire** [comment]
@omnilight are you willing to continue work on this PR?

**Comment by omnilight** [comment]
Yes, shure, I'll work on this PR during current weekend, sorry for delay

**Comment by SilverFire** [comment]
Thank you very much for you effort! :+1:

**Comment by SilverFire** [comment]
@omnilight if you have no time to do it - i'll continue your work

**Comment by omnilight** [comment]
@SilverFire Yes, I do not have time for this now, so it would be greate if you work on this PR, thank you!

**Comment by SilverFire** [comment]
Merged! Thank you! https://github.com/yiisoft/yii2/commit/b68613b7469a7e5df42e0f532cc5370964b07899
https://github.com/yiisoft/yii2/commit/e0f4114d77bef0e3f1847801b3a41f5a9f6dc62e

**Pull request closed without merging**